packages/client/src/base/model/diagram-loader.ts

```diff
@@ -0,0 +1,161 @@
+/********************************************************************************
+ * Copyright (c) 2023 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, multiInject, optional, postConstruct } from 'inversify';
+import {
+    ApplicationIdProvider,
+    Args,
+    EMPTY_ROOT,
+    EndProgressAction,
+    GLSPClient,
+    MaybePromise,
+    RequestModelAction,
+    ServerStatusAction,
+    SetModelAction,
+    StartProgressAction,
+    TYPES
+} from '~glsp-sprotty';
+import { GLSPActionDispatcher } from '../action-dispatcher';
+import { Ranked } from '../ranked';
+
+/**
+ * Instance specific configuration options for a GLSP diagram
+ */
+export interface IDiagramOptions {
+    /**
+     * Unique id associated with this diagram. Used on the server side to identify the
+     * corresponding client session.
+     */
+    clientId: string;
+    /**
+     * The diagram type i.e. diagram language this diagram is associated with.
+     */
+    diagramType: string;
+    /**
+     * The GLSP client used by this diagram to communicate with the server.
+     */
+    glspClient: GLSPClient;
+    /**
+     * The file source URI associated with this diagram.
+     */
+    sourceUri?: string;
+
+    /**
+     * The initial edit mode of diagram. If no defined the default `editable` edit mode will be used
+     */
+    editMode?: string;
+}
+
+/**
+ * Services that implement startup hooks which are invoked during the {@link DiagramLoader.load} process.
+ * Typically used to dispatch additional initial actions and/or activate UI extensions on startup.
+ * Execution order is derived by the `rank` property of the service. If not present, the {@link Ranked.DEFAULT_RANK} will be assumed.
+ *
+ */
+export interface IDiagramStartup extends Partial<Ranked> {
+    /**
+     * Hook for services that should be executed before the underlying GLSP client is configured and the server is initialized.
+     */
+    preInitialize?(): MaybePromise<void>;
+    /**
+     * Hook for services that should be executed before the initial model loading request (i.e. `RequestModelAction`) but
+     * after the underlying GLSP client has been configured and the server is initialized.
+     */
+    preModelLoading?(): MaybePromise<void>;
```

Comment:
I don't have a hard opinion, but how about `preRequestModel?()` and `postRequestModel?()`?
This would be more in line with the startup action names, consistent with `preInitialize`